// docxtemplater/js/filetypes.js
"use strict";

const contentTypes = {
  docx: {
    main: "application/vnd.openxmlformats-officedocument.wordprocessingml.document.main+xml",
    header: "application/vnd.openxmlformats-officedocument.wordprocessingml.header+xml",
    footer: "application/vnd.openxmlformats-officedocument.wordprocessingml.footer+xml",
    footnotes: "application/vnd.openxmlformats-officedocument.wordprocessingml.footnotes+xml",
  },
  docm: {
    main: "application/vnd.ms-word.document.macroEnabled.main+xml",
  },
  dotx: {
    main: "application/vnd.openxmlformats-officedocument.wordprocessingml.template.main+xml",
  },
  dotm: {
    main: "application/vnd.ms-word.template.macroEnabledTemplate.main+xml",
  },
  pptx: {
    slide: "application/vnd.openxmlformats-officedocument.presentationml.slide+xml",
    slideMaster: "application/vnd.openxmlformats-officedocument.presentationml.slideMaster+xml",
    slideLayout: "application/vnd.openxmlformats-officedocument.presentationml.slideLayout+xml",
    presentation: "application/vnd.openxmlformats-officedocument.presentationml.presentation.main+xml",
  },
};

const fileTypes = {
  main: Object.values(contentTypes.docx).concat(Object.values(contentTypes.pptx)),
  docx: Object.values(contentTypes.docx),
  pptx: Object.values(contentTypes.pptx),
};

module.exports = fileTypes;
